models/User.go

```diff
@@ -1677,15 +1793,35 @@ func PerformOrderFillTransaction(ask *Ask, bid *Bid, stockTradePrice uint64, sto
 		bid.IsClosed = oldBidIsClosed
 
 		tx.Rollback()
-		return AskUndone, BidUndone, nil
+	}
+
+	if cashLeft < MINIMUM_CASH_LIMIT {
+		l.Errorf("Check1: Failed. Not enough cash. Returning reserved cash back to user.")
+		if err := bid.Close(tx); err != nil {
+			errorHelper("Unable to close bid. Rolling back. Error: %+v", err)
+			return AskUndone, BidUndone, nil
+		}
+		if err := saveBidCancelOrderTransaction(bid, biddingUser, tx); err != nil {
+			errorHelper("Error saving BidCancelOrderTransaction. Rolling back. Error: %+v", err)
+			return AskUndone, BidUndone, nil
+		}
+		if err := tx.Commit(); err != nil {
```

Comment:
Damn. Nice xD

Reply:
> # Found a bug
> Fix my comment about tx.Commit() before reading any further
> 
> ### Steps to reproduce
> 1. Create 2 users - let's call them user1 and user2
> 2. User1 places **Market** Buy order such that he has no money left after placing the order i.e his cash left should be 0 i.e All his cash has been reserved.
> 3. User2 places Limit Sell order for a price **higher** than the current stock price such that it only **partially** fills the above mentioned Buy order.
> 
> Now, the order will fail since user1 no longer has enough funds to fulfill the order. In which case, you are closing the associated Bid and returning user1's reserved cash. **BUT** you will find that an incorrect amount of reserved cash is given back to the user.
> 
> ### The Reason
> To calculate the reserved cash you are taking into account stockQuantityFulfilled, but unfortunately, in `PerformOrderFillTransaction`, the `stockQuantityFulfilled` is updated **before** you call `saveBidCancelOrderTransaction` (which is responsible for giving back the user's reserved cash). Hence, a less amount is being given back than what is supposed to be.
> 
> ### The Fix
> You need to reset the `stockQuantityFulfilled` value first and then call `saveBidCancelOrderTransaction`
> 
> ### The relevant code
> User.go line 1801
> 
> ```go
> if cashLeft < MINIMUM_CASH_LIMIT {
> 	l.Errorf("Check1: Failed. Not enough cash. Returning reserved cash back to user.")
> 	if err := bid.Close(tx); err != nil {
> 		return errorHelper("Unable to close bid. Rolling back. Error: %+v", AskUndone, BidUndone, err)
> 	}
> 	if err := saveBidCancelOrderTransaction(bid, biddingUser, tx); err != nil {
> 		return errorHelper("Error saving BidCancelOrderTransaction. Rolling back. Error: %+v", AskUndone, BidUndone, err)
> 	}
> 	if err := tx.Commit().Error; err != nil {
> 		return errorHelper("Error while commiting. Rolling back. Error: %+v", AskUndone, BidUndone, err)
> 	}
> 	go updateDataStreams(nil, nil, nil, nil)
> 	go SendNotification(biddingUser.Id, fmt.Sprintf("Your Buy order#%d has been closed due to insufficient cash", bid.Id), false)
> 	return AskUndone, BidDone, nil
> }
> ```
> 
> ### After fixing it
> Continue testing this situation heavily and see what happens i.e The user has placed his Buy order, but then due to a stock price increase, when the order actually gets fulfilled, the reserved cash is still not enough. See if you can find any other bugs.

State will now be reverted before on entering if cashLeft < min